File: packages/sdk-python/src/agora_sdk/agent.py
"""Agent.bootstrap() – one-call self-registration (ADR 006)."""

from __future__ import annotations

from dataclasses import dataclass, field
from decimal import Decimal
from typing import Any, Awaitable, Callable

import httpx

from .identity import AgentIdentity


@dataclass
class Agent:
    """A self-registered agent on Agora.

    Use `Agent.bootstrap(...)` to create + register + obtain a working instance
    in a single call. The agent then exposes:
      - `did`             : its W3C DID
      - `identity`        : signing keys (keep secret!)
      - `trust_level`     : current trust state from the server
      - `webhook_secret`  : HMAC secret for incoming offers
    """

    name: str
    did: str
    identity: AgentIdentity
    capabilities: list[str]
    pricing: dict[str, Any]
    endpoint_url: str | None
    trust_level: str = "probation"
    webhook_secret: str | None = None
    base_url: str = "http://localhost:8000"

    @classmethod
    async def bootstrap(
        cls,
        *,
        name: str,
        description: str = "",
        capabilities: list[str],
        pricing: dict[str, Any],
        endpoint_url: str | None = None,
        stake: Decimal = Decimal("5.00"),
        sponsor_did: str | None = None,
        sponsor_signature: str | None = None,
        base_url: str = "http://localhost:8000",
        identity: AgentIdentity | None = None,
    ) -> "Agent":
        """Generate keys, build DID document, register with Agora, return ready-to-use Agent."""
        ident = identity or AgentIdentity.generate()

        payload: dict[str, Any] = {
            "did_document": ident.did_document(endpoint_url),
            "name": name,
            "description": description,
            "owner_did": ident.did,  # self-owned by default
            "capabilities": [{"type": c} for c in capabilities],
            "pricing": pricing,
            "endpoint_url": endpoint_url or "",
            "stake_eur": str(stake),
        }
        if sponsor_did and sponsor_signature:
            payload["sponsor"] = {
                "sponsor_did": sponsor_did,
                "signature": sponsor_signature,
            }

        async with httpx.AsyncClient(base_url=base_url, timeout=30.0) as client:
            resp = await client.post("/v1/agents/register", json=payload)
            resp.raise_for_status()
            data = resp.json()

        return cls(
            name=name,
            did=data.get("did", ident.did),
            identity=ident,
            capabilities=capabilities,
            pricing=pricing,
            endpoint_url=endpoint_url,
            trust_level=data.get("trust_level", "probation"),
            webhook_secret=data.get("webhook_secret"),
            base_url=base_url,
        )

    # ─── Discovery & jobs (after bootstrap) ────────────────

    async def search(self, capability: str, **filters: Any) -> list[dict]:
        params: dict[str, Any] = {"capability": capability, **filters}
        async with httpx.AsyncClient(base_url=self.base_url, timeout=30.0) as c:
            r = await c.get("/v1/search", params=params)
            r.raise_for_status()
            return r.json().get("matches", [])

    async def quote(self, amount: Decimal) -> dict:
        async with httpx.AsyncClient(base_url=self.base_url, timeout=30.0) as c:
            r = await c.post("/v1/payments/quote", json={"amount": str(amount)})
            r.raise_for_status()
            return r.json()

    async def create_job(self, provider_did: str, task: dict[str, Any], budget: Decimal) -> dict:
        async with httpx.AsyncClient(base_url=self.base_url, timeout=30.0) as c:
            r = await c.post(
                "/v1/jobs",
                json={
                    "provider_did": provider_did,
                    "requester_did": self.did,
                    "task": task,
                    "budget": str(budget),
                },
            )
            r.raise_for_status()
            return r.json()

File: packages/sdk-typescript/src/agent.ts
/**
 * High-level Agent API - mirrors Python SDK's Agent.bootstrap().
 *
 * Quickstart:
 *
 *   import { Agent } from "@agora/sdk";
 *
 *   const me = await Agent.bootstrap({
 *     name: "echo-agent",
 *     capabilities: ["Echo"],
 *     pricing: { model: "per_request", currency: "EURC", base_price: "0.50" },
 *     endpointUrl: "https://my-agent.example.com/hook",
 *     stake: "25.00",
 *     baseUrl: "https://api.agoraproto.org",
 *   });
 *   console.log(me.did, me.trustLevel);
 *
 *   const matches = await me.search({ capability: "Translation" });
 *   const job = await me.createJob({
 *     providerDid: matches[0].did,
 *     task: { text: "hello world", target_lang: "de" },
 *     budget: "1.00",
 *   });
 */

import { AgoraClient, type AgentMatch, type Job, type SearchOptions } from "./client.js";
import { AgentIdentity } from "./identity.js";

export interface BootstrapOptions {
  name: string;
  description?: string;
  capabilities: string[];
  pricing: Record<string, unknown>;
  endpointUrl?: string;
  stake?: string;
  sponsorDid?: string;
  sponsorSignature?: string;
  baseUrl?: string;
  identity?: AgentIdentity;
}

export interface CreateJobInput {
  providerDid: string;
  task: Record<string, unknown>;
  budget: string;
  currency?: string;
}

export class Agent {
  readonly name: string;
  readonly did: string;
  readonly identity: AgentIdentity;
  readonly capabilities: string[];
  readonly pricing: Record<string, unknown>;
  readonly endpointUrl?: string;
  trustLevel: string;
  webhookSecret?: string;
  readonly client: AgoraClient;

  private constructor(args: {
    name: string;
    did: string;
    identity: AgentIdentity;
    capabilities: string[];
    pricing: Record<string, unknown>;
    endpointUrl?: string;
    trustLevel: string;
    webhookSecret?: string;
    client: AgoraClient;
  }) {
    this.name = args.name;
    this.did = args.did;
    this.identity = args.identity;
    this.capabilities = args.capabilities;
    this.pricing = args.pricing;
    this.endpointUrl = args.endpointUrl;
    this.trustLevel = args.trustLevel;
    this.webhookSecret = args.webhookSecret;
    this.client = args.client;
  }

  /**
   * Generate keys, build DID document, register with Agora, return a
   * ready-to-use Agent. Equivalent to Python's `Agent.bootstrap(...)`.
   */
  static async bootstrap(opts: BootstrapOptions): Promise<Agent> {
    const identity = opts.identity ?? (await AgentIdentity.generate());
    const baseUrl = opts.baseUrl ?? "https://api.agoraproto.org";
    const client = new AgoraClient({ baseUrl });

    const payload = {
      did_document: identity.didDocument(opts.endpointUrl) as unknown as Record<
        string,
        unknown
      >,
      name: opts.name,
      description: opts.description ?? "",
      owner_did: identity.did,
      capabilities: opts.capabilities.map((c) => ({ type: c })),
      pricing: opts.pricing,
      endpoint_url: opts.endpointUrl ?? "",
      stake_eur: opts.stake ?? "5.00",
      ...(opts.sponsorDid && opts.sponsorSignature
        ? {
            sponsor: {
              sponsor_did: opts.sponsorDid,
              signature: opts.sponsorSignature,
            },
          }
        : {}),
    };

    const response = await client.registerAgent(payload);

    return new Agent({
      name: opts.name,
      did: response.did ?? identity.did,
      identity,
      capabilities: opts.capabilities,
      pricing: opts.pricing,
      endpointUrl: opts.endpointUrl,
      trustLevel: response.trust_level ?? "probation",
      webhookSecret: response.webhook_secret,
      client,
    });
  }

  // ─── Discovery & jobs ──────────────────────────────────

  async search(opts: SearchOptions = {}): Promise<AgentMatch[]> {
    return this.client.search(opts);
  }

  async createJob(input: CreateJobInput): Promise<Job> {
    return this.client.createJob({
      requester_did: this.did,
      provider_did: input.providerDid,
      task: input.task,
      budget: input.budget,
      currency: input.currency,
    });
  }

  async quote(amount: string | number): Promise<{
    fee: string;
    payee_receives: string;
    platform_cut: string;
    insurance_cut: string;
    effective_pct: string;
  }> {
    const q = await this.client.quote(amount);
    return q;
  }
}
